Add a description of the proposed owners file system for the repo

brendandburns opened the pull request:
@bgrant0607 @davidopp @thockin @dchen1107 @roberthbailey

brendandburns committed:
Add a description of the proposed owners file system for the repo

roberthbailey commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,72 @@
+
+## Overview
+We want to establish owners for different parts of the code in the Kubernetes codebase.  These owners
+will serve as the approvers for code to be submitted to these parts of the repository.  Notable, owners
```
nit: Notably

brendandburns replied:
done.

roberthbailey commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,72 @@
+
+### Step Two: A PR is LGTM'd
+Once a PR is reviewed and LGTM'd it is eligible for submission.  However, in order for it to be submitted
+all owners have to 'approve' the PR.  Approval comes in two forms:
```
'all owners' is a little vague. I think you mean all owners listed by the bot above, which is really just owners over the parts of the repository that are being changed.

brendandburns replied:
Updated to clarify.  I wanted to distinguish between multiple owners for a file (where one of N approver is required) and across multiple files (which may require multiple OWNERS)

roberthbailey commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,72 @@
+Github has the ability to [require status checks prior to merging](https://help.github.com/articles/enabling-required-status-checks/)
+
+Once we have the status check munger described above implemented, we will add this required status check
+to our main branch as well as any release branches.
```
How will we handle owners on release branches? Will we read the owners files from the branch or from master? Or have separate branch owners?

(I think that the bot doesn't do a good job today of distinguishing cherry picks from master merges)

brendandburns replied:
read OWNERS from release branch, at least to start...

eparis replied:
yuck (from the bot PoV) but we can certainly find a way.

davidopp commented on docs/devel/owners.md:
````diff
@@ -0,0 +1,72 @@
+This comment will say:
+
+```
+Approval is required from <owner-a>, <owner-b>
````
I assume the comma means "or" -- so better to spell it out

davidopp replied:
Oh, looks like you meant "and" (later you say "all owners have to approve") -- is this what we want? What if an owner is on an extended vacation? I guess we could say an owner should remove themselves from the owners file if they are going to be on an extended vacation. But then (1) the annotation for any existing PRs will be wrong (it will mention them when they are not actually required) and (2) once they return the annotation for PRs submitted while they were out will be wrong (it will not mention them but they are actually required). Of course I'm assuming the mergebot uses the current owners file, not whatever was recorded in the PR thread (this seems like the only reasonable policy). 

The easiest solution is to make it an "or" -- though I can see downsides to that too. But those at least could be solved by social agreement (for example, if A and B are owners, they could implicitly implement "and" as follows: they agree that normally neither will write "I approve" until both have "LGTMd" - the second to "LGTM" will write "I approve". But if one is on vacation, then the other could write "I approve" without an LGTM from the one on vacation.)

So I would suggest "or." I'm assuming we will have a small enough group of owners for each part of the codebase (e.g. 2-3 people) that such social agreements will be workable.

brendandburns replied:
I clarified a bunch of this, hopefully it makes more sense now.  I didn't mean all owners, I just meant that a PR could require multiple owners due to cross multiple different directories.

davidopp commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,72 @@
+we need to add owners files to the repository.
+
+### Approval Munger
+We need to add a munger that adds comments to PRs indicating whos approval they require.  This munger will
```
s/whos/whose/

brendandburns replied:
done.

davidopp commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,72 @@
+## High Level flow
+
+### Step One: A PR is submitted
+After a PR is submitted, a user can request the approval from the required owners.  The user requests this
```
Why not just do this automatically on every PR when it is submitted?

brendandburns replied:
There were some complains about the bot spamming PRs (and thus email) trying to balance things here, I'm ok w/ auto sending on every PR also...

davidopp commented:
Great proposal. My only comment is about "or" vs. "and"; see inline comment.

brendandburns commented:
Comments addressed, ptal.

Thanks!
--brendan

roberthbailey commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,126 @@
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/devel/owners.md).
```
This is a new file so this will be a dead link until we cut a new release that includes this doc.

brendandburns replied:
this is auto-generated, I'm not sure I can remove it w/o the tools complaining...

roberthbailey replied:
Can we file an issue to fix the tools? This is a problem any time we add a new file to the repo.

thockin replied:
It's also a problem if we move or rename files.

roberthbailey commented on docs/devel/owners.md:
````diff
@@ -0,0 +1,126 @@
+by writing:
+
+```
+@k8s-bot approvers
````
I agree with @davidopp -- can we just do this automatically for PRs that are LGTM'd but don't have the requires approvals?

brendandburns replied:
done.

davidopp commented on docs/devel/owners.md:
````diff
@@ -0,0 +1,126 @@
+This comment will say:
+
+```
+Approval is required from <owner-a> OR <owner-b> AND <owner-c> OR <owner-d> ...
````
putting a comma before each AND would make this clearer, and be easy to implement :)

brendandburns replied:
done.

davidopp commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,126 @@
+approval from more than one person, this is where the `AND` clauses come from.
+
+PR authors are not required to ask for approval, if for example, the reviewer is already an owner, it is
+expected that the reviewer will (eventually) grant approval without being explicitly asked to do so.
```
This sentence is written in a way that makes the meaning unclear. I'd suggest "There are some circumstances under which the PR author is not required to ask for approval. For example, if the PR author is an owner in all modified directories, or if the reviewer is an owner."

But I think it is easier to just automatically add the "Approval is required from..." line always, instead of requiring people to figure out whether they need to add "@k8s-bot approvers" , etc.

brendandburns replied:
done.

brendandburns commented:
ok, I made the updates automatic.  if people complain about spam, I'm sending them to you :P

ptal.

roberthbailey commented:
lgtm

bgrant0607 commented:
cc @eparis @smarterclayton @derekwaynecarr @ncdc

bgrant0607 commented:
Thanks. Blunderbuss should take owners into account, and we should rotate automatic owner approval notification as well.

bgrant0607 commented:
Once this is implemented, approval will need to be added to our review flow documentation.  #13935
cc @sarahnovotny

smarterclayton commented:
Proposal is LGTM


On Nov 17, 2015, at 6:22 PM, Brian Grant <notifications@github.com> wrote:

cc @eparis <https://github.com/eparis> @smarterclayton
<https://github.com/smarterclayton> @derekwaynecarr
<https://github.com/derekwaynecarr> @ncdc <https://github.com/ncdc>

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17267#issuecomment-157542708>
.

ncdc commented:
LGTM

eparis commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,128 @@
+### Status Munger
+
+GitHub has a [status api](https://developer.github.com/v3/repos/statuses/), we will add a status munger that pushes a status onto a PR of approval status.  This status will only be approved if the relevant
+approvers have approved the PR.
```
https://github.com/kubernetes/contrib/pull/252

eparis commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,128 @@
+Github has the ability to [require status checks prior to merging](https://help.github.com/articles/enabling-required-status-checks/)
+
+Once we have the status check munger described above implemented, we will add this required status check
+to our main branch as well as any release branches.
```
I see little purpose in this. The bot won't care (since it will both be setting the status and doing the merge only if the status would be set) and if humans are doing things wrong we need to 'discuss' with the humans. When the build breaks and we have a 1-liner to fix it I WANT a human to be able to just push it.

If we want/need a munger which flags those cases where someone merged something by hand, we can do that, but I'd really like to not prevent people from doing useful things in a timely manor (and slapping their wrists if they mess up)

davidopp replied:
@eparis I'm not sure there's a way to satisfy what you want and what this proposal is trying to do. We specifically don't want all committers to be able to hit the "big green button" on any part of the codebase. Maybe a way to deal with the emergency situation you described (which I think is a legitimate point), is to ensure that the oncall has OWNER permission for the entire codebase for the entire duration of their oncall shift, so there's always someone oncall to merge emergency stuff.

eparis replied:
Is the problem people hitting the big green button or people LGTM things which they maybe didn't know enough to LGTM? Today the big green button and LGTM are basically the same. I feel like the 'big green button' we should solve socially, and the latter we should solve with OWNERS.

We can create a report/munger which tracks the PRs where people hit the 'big green button' which could occasionally/regularly be reviewed. If someone regularly is hitting the merge button and shouldn't be, they don't deserve write/merge/push privs....

davidopp replied:
> Is the problem people hitting the big green button or people LGTM things which they maybe didn't know enough to LGTM? Today the big green button and LGTM are basically the same. I feel like the 'big green button' we should solve socially, and the latter we should solve with OWNERS.

Well, it addresses both.

But I think I understand your proposal: OWNERS should apply to adding LGTM label, but all committers should be able to hit "big green button" but that should be logged and audited.

That doesn't sound particularly unreasonable to me assuming the auditing facility is reliable and secure. But I'm not sure what Brendan or others think. (Or if there's some loophole I'm not considering.)

eparis replied:
Just to put numbers to the discussion. `git log --merges --format="%an" v1.1.0..HEAD | sort | uniq -c` This is all of the people who clicked the big green button between 1.1.0 and HEAD. The people at the bottom are the 'owners' who I assume would be exempt from this enforcement.

I'm not sure what these are telling us. But maybe it is telling us we need to start the social engineering now   :)

```
     12 Abhi Shah
     55 Alex Robinson
      3 CJ Cullen
      2 David Eads
      1 Derek Carr
      2 Dr. Stefan Schimanski
      5 Fabio Yeon
      4 Isaac Hollander McCreery
      4 Janet Kuo
      1 Jeff Grafton
     19 Jeff Lowdermilk
     14 Jerzy Szczepkowski
      3 Karl Isenberg
      3 krousey
      9 Marcin Wielgus
     46 Marek Grabowski
      8 Nikhil Jindal
      3 Paul Morie
      2 Phillip Wittrock
     26 Piotr Szczesniak
      4 Prashanth B
      2 Quinton Hoole
      9 Robert Bailey
      9 Saad Ali
      6 Vincenzo D'Amore
     21 Wojciech Tyczynski
      1 Yu-Ju Hong
      -----------------
    274 Total

    671 k8s-merge-robot

      1 * Alex Mohr
     42 * Brendan Burns
      7 * Brian Grant
     16 * Daniel Smith
      5 * David Oppenheimer
     53 * Dawn Chen
      9 * Eric Tune
     22 * Filip Grzadkowski
     23 * Mike Danese
     11 * Tim Hockin
      2 * Zach Loafman
      ------------------
    191 Total
```

brendandburns replied:
@eparis nothing in this prevents a maintainer from hitting the green button to merge a qualified PR.

However it _does_ require that the PR be approved by an OWNER.  This could be the author of the PR, if they are an OWNER for that area, or they will need to find someone.  That's explicitly the point of this PR.  I don't think committers should be able to commit code without OWNER approval.

(and obviously we will have some top-level OWNERS for emergencies)

davidopp commented:
> Blunderbuss should take owners into account, and we should rotate automatic owner approval notification as well.

I didn't understand this comment. I thought the idea was that every PR gets a message at submission time listing all potential owners who could/must approve, so the submitter knows everyone who can approve.

Is the concern that if the tool @ mentions all potential owners, then people who are owners will get a lot of spam (especially since there is no way to un-Cc yourself from an issue/PR)? I think that's a very valid point -- my very strong suggestion would be that the message that lists who are the owners, should **not** list them using @ format, but just by username. (For example "this PR must be approved by brendandburns OR davidopp" rather than "this PR must be approved by @brendandburns or @kubernetes/rh-dev-experience") Then I think it's fine to list everyone.

eparis commented:
#1 OWNERS should replace https://github.com/kubernetes/contrib/blob/master/mungegithub/blunderbuss.yml

#2 what if we don't cc/spam OWNERS (using @) until after LGTM?  So yes, all owners will get spam on all PRs in their area, but only those someone has LGTM'd....

davidopp commented on docs/devel/owners.md:
````diff
@@ -0,0 +1,128 @@
+This comment will say:
+
+```
+Approval is required from <owner-a> OR <owner-b>, AND <owner-c> OR <owner-d>, AND ...
````
I think you should make it clear that these owners will be listed without the @ symbol in front of their name. Otherwise every owner is going to be irrevocably Cc'd on every PR in their area, which will drive them insane.

brendandburns replied:
done.

davidopp commented on docs/devel/owners.md:
````diff
@@ -0,0 +1,128 @@
+Approval is required from <owner-a> OR <owner-b>, AND <owner-c> OR <owner-d>, AND ...
+```
+
+The set of required authors is drawn from the OWNERS files in the repository (see below).  For each file
````
s/authors/owners/

brendandburns replied:
done.

davidopp commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,128 @@
+
+### Step Three: A PR is merged
+
+Once a PR is LGTM'd and all owners have approved, it is eligible for merge.  The merge bot takes care of
```
s/all owners/all required owners/

brendandburns replied:
done.

thockin commented on docs/devel/owners.md:
```diff
@@ -0,0 +1,128 @@
+
+Owner approval comes in two forms:
+
+   * An owner adds a comment to the PR saying "I approve"
```
Why not LGTM from the owner?

eparis replied:
My first thought was a `LGTM-owner` label, but changes which spam areas will need multiple owner approvals and the only way I know of to do that is in comments.

thockin replied:
I meant "LGTM" as comment rather than "I approve"

On Tue, Nov 17, 2015 at 9:38 PM, Eric Paris <notifications@github.com>
wrote:

> In docs/devel/owners.md
> <https://github.com/kubernetes/kubernetes/pull/17267#discussion_r45161784>
> :
>
> > +```
> > +
> > +The set of required authors is drawn from the OWNERS files in the repository (see below).  For each file
> > +there should be multiple different OWNERS, these owners are listed in the `OR` clause(s). Because
> > +it is possible that a PR may cover different directories, with disjoint sets of OWNERS, a PR may require
> > +approval from more than one person, this is where the `AND` clauses come from.
> > +
> > +### Step Two: A PR is LGTM'd
> > +
> > +Once a PR is reviewed and LGTM'd it is eligible for submission.  However, for it to be submitted
> > +an owner for all of the files changed in the PR have to 'approve' the PR.  A user is an owner for a
> > +file if they are included in the OWNERS hierarchy (see below) for that file.
> > +
> > +Owner approval comes in two forms:
> > +
> > +   * An owner adds a comment to the PR saying "I approve"
>
> My first thought was a LGTM-owner label, but changes which spam areas
> will need multiple owner approvals and the only way I know of to do that is
> in comments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17267/files#r45161784>.
>

davidopp replied:
Only counter-argument would be that LGTM implies you reviewed and understand the code whereas "I approve" might just mean "I trust the person who did the review" or "I verified this code is doing nothing horrible, but I don't really know if it's correct."

thockin replied:
fair enough, I can buy that.  I'd prefer "approved" to "I approve", but
it's noise.

On Tue, Nov 17, 2015 at 9:48 PM, David Oppenheimer <notifications@github.com
> wrote:

> In docs/devel/owners.md
> <https://github.com/kubernetes/kubernetes/pull/17267#discussion_r45162120>
> :
>
> > +```
> > +
> > +The set of required authors is drawn from the OWNERS files in the repository (see below).  For each file
> > +there should be multiple different OWNERS, these owners are listed in the `OR` clause(s). Because
> > +it is possible that a PR may cover different directories, with disjoint sets of OWNERS, a PR may require
> > +approval from more than one person, this is where the `AND` clauses come from.
> > +
> > +### Step Two: A PR is LGTM'd
> > +
> > +Once a PR is reviewed and LGTM'd it is eligible for submission.  However, for it to be submitted
> > +an owner for all of the files changed in the PR have to 'approve' the PR.  A user is an owner for a
> > +file if they are included in the OWNERS hierarchy (see below) for that file.
> > +
> > +Owner approval comes in two forms:
> > +
> > +   * An owner adds a comment to the PR saying "I approve"
>
> Only counter-argument would be that LGTM implies you reviewed and
> understand the code whereas "I approve" might just mean "I trust the person
> who did the review" or "I verified this code is doing nothing horrible, but
> I don't really know if it's correct."
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17267/files#r45162120>.
>

brendandburns replied:
what David said, LGTM means I like the code.  'I approve' means its ok to merge this.

davidopp commented:
> #1 OWNERS should replace https://github.com/kubernetes/contrib/blob/master/mungegithub/blunderbuss.yml

I don't understand how it's a replacement. Blunderbuss assigns reviewers (and probably other stuff I've forgotten); I thought reviewers and OWNERS are intentionally orthogonal.

> #2 what if we don't cc/spam OWNERS (using @) until after LGTM? So yes, all owners will get spam on all PRs in their area, but only those someone has LGTM'd....

That might be a reasonable compromise, though I can imagine someone still wouldn't want so much email (like me :) ). I guess we could make the behavior customizable, so the OWNERS file could specify \<owner, preference\> tuples where the preference is a bool that says whether the owner gets @ mentioned or not after LGTM....

eparis commented:
The blunderbuss picks an assignee from the list of people kept entirely outside of the tree. I was imagining we could just pick anyone from OWNERS to be the assignee. I think we'd get better assignee lists if it wasn't kept hidden off in some other repo. Although the number of people qualified to review the PR is likely greater than the number of owners.

Is there any interest in making the owners file structured in some way? So we can list both 'reviewers' and 'approvers' in the file?  Blunderbuss would take into account the 'reviewers'? It's not really this PR, but I'd like to see the reviewers list most into the tree rather than be out here in the YAML....

ncdc commented:
+1 to in-tree. It makes the involved/responsible parties more visible to the community.

smarterclayton commented:
For the "oh no oh no" case any of the GitHub owners can still do merges.  I
don't know that we've had a problem yet where the "oh no oh no" wasn't
something that a direct merge couldn't solve.

On Wed, Nov 18, 2015 at 10:23 AM, Andy Goldstein <notifications@github.com>
wrote:

> +1 to in-tree. It makes the involved/responsible parties more visible to
> the community.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17267#issuecomment-157746850>
> .
>

brendandburns commented:
I will point out that blunderbuss originally _was_ in the tree :P

anyway, I'm fine with adding REVIEWERS files along with the OWNERS files, but I think its beyond the scope of this document.

I have addressed all comments, I'd like to get this merged so someone can start building.

Any last minute feedback?

Thanks!
--brendan

eparis commented:
agree, blunderbuss merger out of scope.

only point of contention is 'eventual' `Requiring approval status`   I don't like that part unless there is a reason.

eparis commented:
(maybe the mere fact there have been 465 'out of process' merges since 1.1.0 is enough reason, but when 40% of those won't be affected...)

davidopp commented:
LGTM. I'll wait until @bgrant0607 gets back to his desk to see if he has any objections before adding the label though...

davidopp commented:
(the PR was auto-assigned to him)

ArtfulCoder commented:
manually merging since k8s-bot is blocked on scalability test failure

ArtfulCoder commented:
I just saw David's comment. I saw the label, but now, we can add changes to the file if there is feedback from Brian.

bgrant0607 commented:
Please don't block on me. We can make it better over time.

derekwaynecarr commented:
I am late to the part, but this looks LGTM to me.

eparis commented:
https://github.com/kubernetes/contrib/pull/282 is a very early first stab WIP. It needs lots more work to be useful and to meet these goals/requirements, but right now it will look at OWNERS files and set/unset a label depending on if they were met.
